cargo-optee/src/ta_builder.rs

```diff
@@ -0,0 +1,505 @@
+
+    // Determine target and cross-compile based on arch
+    let (target, cross_compile) = get_target_and_cross_compile(config.arch, config.std)?;
+
+    // Setup build command with common environment (without --manifest-path)
+    let (mut build_cmd, _temp_dir) = setup_build_command(config, "build")?;
+
+    if !config.debug {
+        build_cmd.arg("--release");
+    }
+
+    // Configure linker
+    let linker = format!("{}gcc", cross_compile);
+    let linker_cfg = format!("target.{}.linker=\"{}\"", target, linker);
+    build_cmd.arg("--config").arg(&linker_cfg);
+
+    // Print the full cargo build command for debugging
+    print_cargo_command(&build_cmd, "Building TA binary");
+
+    let build_output = build_cmd.output();
+
+    // Restore original directory before checking results
+    std::env::set_current_dir(original_dir)?;
+
+    let build_output = build_output?;
+    if !build_output.status.success() {
+        print_output_and_bail("build", &build_output)?;
+    }
+
+    Ok(())
+}
+
+fn get_package_name(manifest_path: &Path) -> Result<String> {
+    if !manifest_path.exists() {
+        bail!("Cargo.toml not found at: {:?}", manifest_path);
+    }
+
+    let cargo_toml_content = fs::read_to_string(manifest_path)?;
+    let cargo_toml: Value = toml::from_str(&cargo_toml_content)?;
+
+    let package_name = cargo_toml
+        .get("package")
+        .and_then(|p| p.get("name"))
+        .and_then(|n| n.as_str())
+        .ok_or_else(|| anyhow::anyhow!("Could not find package name in Cargo.toml"))?;
+
+    Ok(package_name.to_string())
+}
+
+fn strip_binary(config: &TaBuildConfig, manifest_path: &Path) -> Result<(PathBuf, PathBuf)> {
+    println!("Stripping binary...");
+
+    // Determine target based on arch
+    let (target, cross_compile) = get_target_and_cross_compile(config.arch, config.std)?;
+
+    let profile = if config.debug { "debug" } else { "release" };
+
+    // Get the actual package name from Cargo.toml
+    let package_name = get_package_name(manifest_path)?;
+
+    // Use cargo metadata to get the target directory that cargo is actually using
+    let output = Command::new("cargo")
+        .arg("metadata")
+        .arg("--manifest-path")
+        .arg(manifest_path)
+        .arg("--format-version")
+        .arg("1")
+        .arg("--no-deps")
+        .output()?;
+
+    if !output.status.success() {
+        bail!("Failed to get cargo metadata");
+    }
+
+    let metadata: serde_json::Value = serde_json::from_slice(&output.stdout)?;
+    let target_directory = metadata
+        .get("target_directory")
+        .and_then(|v| v.as_str())
+        .ok_or_else(|| anyhow::anyhow!("Could not get target directory from cargo metadata"))?;
+
+    let target_dir = PathBuf::from(target_directory);
+    let profile_dir = target_dir.join(target).join(profile);
+    let binary_path = profile_dir.join(&package_name);
+
+    if !binary_path.exists() {
+        bail!("Binary not found at {:?}", binary_path);
+    }
```

Comment:
All the same pattern can move into a fn like `join_and_check()`, using that fn avoids we forget  check the path